mne/decoding/ssd.py

```diff
@@ -290,3 +301,37 @@ def apply(self, X):
         pick_patterns = self.patterns_[self.sorter_spec][:self.n_components].T
         X = pick_patterns @ X_ssd
         return X
+
+
+def _dimensionality_reduction(cov_signal, cov_noise, info, rank):
+    """Perform dimensionality reduction on the covariance matrices."""
+    from scipy import linalg
+    n_channels = cov_signal.shape[0]
+    rank_signal = list(compute_rank(
+        Covariance(cov_signal, info.ch_names, list(), list(), 0,
+                   verbose=_verbose_safe_false()),
+        rank, _handle_default('scalings_cov_rank', None), info).values())[0]
+    rank_noise = list(compute_rank(
+        Covariance(cov_signal, info.ch_names, list(), list(), 0,
+                   verbose=_verbose_safe_false()),
+        rank, _handle_default('scalings_cov_rank', None), info).values())[0]
+    rank = np.min([rank_signal, rank_noise])  # should be identical
+
+    if rank < n_channels:
+        eigvals, eigvects = linalg.eigh(cov_signal)
+        # sort in descending order
+        ix = np.argsort(eigvals)[::-1]
+        eigvals = eigvals[ix]
+        eigvects = eigvects[:, ix]
+        # compute dimensionality reduction transformation matrix
+        dim_red = np.matmul(
+            eigvects[:, :rank], np.eye(rank) * (eigvals[:rank]**-0.5))
+        logger.info(
+            "Reducing covariance rank from %i -> %i" % (n_channels, rank,))
```

Comment:
The rank of the covariance is not reduced, that is given by your matrix itself. What it is being done here is more a subspace projection. I would recommend changing the info msg.

Reply:
I am not sure I understand entirely. Applying `dim_red` to the covariance matrices later in the function does reduce their rank, right?
https://github.com/mne-tools/mne-python/blob/ae7ecf89f6d531e6b52ae69ea445be4a449de5e5/mne/decoding/ssd.py#L335-L336

Reply:
well, you are estimating the rank of the two matrices and taking the smallest value (which BTW, is there any practical case in which the `rank(cov_signal) != rank(cov_noise)` ? )

So, personally, I don't see this step as a dimensionality reduction of "the rank" (as you have stated in the log message) but a projection to the rank subspace of the data. In these lines, you are in fact making a projection. so dim_reg is a projection matrix to the rank subspace.

Reply:
> BTW, is there any practical case in which the rank(cov_signal) != rank(cov_noise) ?

I could not think of anything, but maybe there is some super strange edge case where the ranks are not equal. Since `np.min` of two values is so computationally cheap, I thought it could be better safe than sorry to check.

> So, personally, I don't see this step as a dimensionality reduction of "the rank" (as you have stated in the log message) but a projection to the rank subspace of the data. In these lines, you are in fact making a projection. so dim_reg is a projection matrix to the rank subspace.

Okay, I think I understand. So something like "Projecting covariance of X channels to Y rank subspace" would be more appropriate?